proposals/NNNN-System-Request.md

```diff
@@ -0,0 +1,94 @@
++           SDL shall send decoded data to the mobile app in hybrid part of message
++        </description>
++   </param>
+</function>
+
+
+<function name="SystemRequest" messagetype="request">
+    ...
++   <param name="requestData" type="String" maxlength="65535" mandatory="false">
++       <description>
++         If it exists, "fileName" shall be ignored;
++         If the length of incoming binary data from hybrid message is less than configured bytes, instead of saving it to a binary file, SDL base64 encoded the data.  
++       </description>
++   </param>
+</function>
+
+
+<function name="SystemRequest" messagetype="response">
+    ...
++   <param name="resultData" type="String" maxlength="65535" mandatory="false">
++       <description> Base64 encoded result data from the system to SDL. SDL shall send decoded data to the mobile app in hybrid part of message. </description>
++   </param>
+</function>
+```
+
+
+Here are the changes needed by SDL core.
+- Add a new configurable parameter in smartdevicelink.ini
+
+```
+; The payload in SystemRequest/OnSystemRequest packet bigger than this value will be saved to a file when passing to HMI,
+; payload smaller than this value will be based64 encoded and stored as a string parameter to HMI 
+; if not specified, this value will default to 0
+MaximumBinaryPayloadSizeToString = 11520
+```
+
+- When core receives a SystemRequest request from mobile, if the incoming binary data is less than a certain size specified by smartdevicelink.ini, instead of saving the data to a temporary file, SDL shall encode the data using base64 format, and send the encoded string directly within a SystemRequest HMI message. If the size of data is large, SDL saves the data to a file and pass the file name to HMI as before.
+
+- When core revives a SystemRequest response or OnSystemRequest notification from HMI, if the optional `data`/`resultData` parameter exists, SDL shall ignore the `fileName` parameter. Instead of reading data from the file, SDL shall base64 decode the string, and send the resulting data in hybrid part of the message to mobile. If the optional `data`/`resultData` parameter does not exist, SDL read data from file as before.
+
+Potential changes to HMI
+
+- When HMI receives a SystemRequest request, if optional parameter `requestData` exists, it shall ignore the `fileName` parameter. 
+- When HMI sends a SystemRequest response or a OnSystemRequest notification, it is free to choose either the file transfer (as before) or string transfer (new) regardless of the configurable parameter in smartdevicelink.ini.
+
+Potential changes to the mobile proxy lib
+
+- Extract hybrid part of the message from SystemRequest RPC response message, if the library does not do that before.
+
+
+
+## Potential downsides
+
+Instead of writing/reading from a file, SDL needs to do base64 encoding/decoding. The length ratio of base64 encoded bytes to input bytes is about 4:3 (33% overhead). 
```

Comment:
```suggestion
Instead of writing/reading from a file, SDL Core needs to do base64 encoding/decoding. The length ratio of base64 encoded bytes to input bytes is about 4:3 (33% overhead). 
```

Reply:
updated